File: src/main/resources/net/contextfw/demo/web/components/ProgressHolder.js
var ProgressHolder = Component.extend({
  init : function(id) {
    this._super(id);
    var self = this;
    this.bindEvents();
  },
  bindEvents : function() {
    var self = this;
    this.el("add").click(function() {
      self.call("addProgress")(self.el("name").val(), self.el("duration").val());
    });
  }
});

var ProgressIndicator = Component.extend({
  init : function(id, progress) {
    this._super(id);
    this.bindEvents();
    this.updateProgress(progress);
  },
  bindEvents : function() {
  },
  updateProgress : function(progress) {
    this.el("progressbar").progressbar({
      value: progress*100
    });
  }
});

File: src/main/java/net/contextfw/demo/web/components/ProgressHolder.java
package net.contextfw.demo.web.components;

import java.util.ArrayList;
import java.util.List;

import net.contextfw.demo.services.CommonService;
import net.contextfw.web.application.component.Element;
import net.contextfw.web.application.remote.Remoted;
import net.contextfw.web.application.scope.Provided;

import com.google.inject.Inject;

public class ProgressHolder extends JsComponent {

    @Inject
    @Provided
    private CommonService commonService;
    
    @Element
    private List<ProgressIndicator> indicators = new ArrayList<ProgressIndicator>();

    @Remoted
    public void addProgress(String name, long duration) {
        if (indicators.size() < 10 && duration > 0 && duration <= 30) {
            long now = System.currentTimeMillis();
            ProgressIndicator indicator = registerChild(new ProgressIndicator(name));
            indicators.add(indicator);
            commonService.executeAsync(new Job(indicator, now, duration*1000), 0);
            refresh();
        }
    }
    
}
